### PR TITLE
Fix NoSuchElementException when rereplicate empty ledgers

### PR DESCRIPTION
**Pull request by chenhongSZ**
### Motivation
Master issue: #4036 

### Changes
Set the `numberOfEntriesToReplicate` to 0 when ledger is empty

### CODE REVIEW

**Review by aloyszhang — APPROVED** [review]

**Comment by aloyszhang** [comment]
@hangc0276 @zymap PTAL

**Comment by chenhongSZ** [comment]
> Nice catch. @chenhongSZ Would you please add a test to cover this change? thanks.

OK I'll finish it next week

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerFragmentReplicator.java [hunk 1]
```diff
@@ -293,7 +293,8 @@ static Set<LedgerFragment> splitIntoSubFragments(LedgerHandle lh,
             assert false;
         }
 
-        long numberOfEntriesToReplicate = (lastEntryId - firstEntryId) + 1;
+        long numberOfEntriesToReplicate = firstEntryId == INVALID_ENTRY_ID && lastEntryId == INVALID_ENTRY_ID ? 0 :
+                (lastEntryId - firstEntryId) + 1;
         long splitsWithFullEntries = numberOfEntriesToReplicate
```
Comment (wenbingshen) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerFragmentReplicator.java:298]: Can you share the configuration of `replicationEntryBatchSize` in the bookie server?
If I understand correctly, this problem will only occur when `replicationEntryBatchSize=1`

Comment (chenhongSZ) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerFragmentReplicator.java:298]: Your understanding is correct. Our configuration is `rereplicationEntryBatchSize=1`

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerFragmentReplicator.java [hunk 2]
```diff
@@ -293,7 +293,8 @@ static Set<LedgerFragment> splitIntoSubFragments(LedgerHandle lh,
             assert false;
         }
 
-        long numberOfEntriesToReplicate = (lastEntryId - firstEntryId) + 1;
+        long numberOfEntriesToReplicate = firstEntryId == INVALID_ENTRY_ID && lastEntryId == INVALID_ENTRY_ID ? 0 :
```
Comment (hangc0276) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerFragmentReplicator.java:296]: We can simplify the condition to `long numberOfEntriesToReplicate = firstEntryId == INVALID_ENTRY_ID ? 0 : (lastEntryId - firstEntryId) + 1;`

**Review by hangc0276 — APPROVED** [review]

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerFragmentReplicator.java (hunk 2 above)
Comment (chenhongSZ) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerFragmentReplicator.java]: done, PTAL

Comment (hangc0276) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerFragmentReplicator.java]: LGTM

**Review by aloyszhang — APPROVED** [review]

**Review by wenbingshen — APPROVED** [review]

**Pull request merged**